packages/smooth_app/lib/pages/product_copy_helper.dart

```diff
@@ -0,0 +1,172 @@
+    }
+    // TODO(monsieurtanuki): add "add product list / pantry / shopping list" buttons
+    final List<PantryType> pantryTypes = <PantryType>[
+      PantryType.PANTRY,
+      PantryType.SHOPPING,
+    ];
+    for (final PantryType pantryType in pantryTypes) {
+      final List<Pantry> pantries = allPantries[pantryType];
+      int index = 0;
+      for (final Pantry pantry in pantries) {
+        if (ignoredPantry != null &&
+            ignoredPantry.name == pantry.name &&
+            ignoredPantry.pantryType == pantry.pantryType) {
+          // skip
+        } else {
+          children.add(
+            PantryButton(
+              pantries,
+              index,
+              onPressed: () => Navigator.pop<Pantry>(context, pantry),
+            ),
+          );
+        }
+        index++;
+      }
+    }
+    return children;
+  }
+
+  Future<void> copy({
+    @required final BuildContext context,
+    @required dynamic target,
+    @required final Map<PantryType, List<Pantry>> allPantries,
+    @required final DaoProductList daoProductList,
+    @required final List<Product> products,
+    @required final UserPreferences userPreferences,
+  }) async {
+    int count; // late
+    Widget Function(BuildContext) go; // late
+    if (target is ProductList) {
+      count = await _addToProductList(
+        daoProductList,
+        target,
+        products,
+      );
+      go = (BuildContext context) => ProductListPage(target);
+    } else if (target is Pantry) {
+      final List<Pantry> pantries = allPantries[target.pantryType];
+      int index = 0;
+      for (final Pantry pantry in pantries) {
+        if (pantry.name == target.name) {
+          final int value = index;
+          count = await _addToPantry(
+            pantries,
+            value,
+            target.pantryType,
+            products,
+            userPreferences,
+          );
+          go = (BuildContext context) => PantryPage(
+                pantries,
+                value,
+                target.pantryType,
+              );
+        }
+        index++;
+      }
+    } else {
+      throw Exception('unknown type $target');
+    }
+    ScaffoldMessenger.of(context).showSnackBar(
+      SnackBar(
+        content: Text('$count products actually added'),
+        duration: const Duration(seconds: 5),
+        action: SnackBarAction(
+          label: 'GO',
+          onPressed: () {
+            Navigator.pop(context);
+            Navigator.push<Widget>(
+              context,
+              MaterialPageRoute<Widget>(builder: go),
+            );
+          },
+        ),
+      ),
+    );
+  }
+
+  Future<int> _addToProductList(
+    final DaoProductList daoProductList,
+    final ProductList target,
+    final List<Product> products,
+  ) async {
+    await daoProductList.get(target);
+    int count = 0;
+    for (final Product product in products) {
+      if (target.barcodes.contains(product.barcode)) {
+        // do nothing?
+      } else {
```

Comment:
```suggestion
      if (!target.barcodes.contains(product.barcode)) {
```

Reply:
I understand your point.
The thing is that I want to make that `if` statement obvious: it's not some kind of "bad luck" or dismissible case, it's an interesting functional point that may need attention. What should we do when that happens? Even you suggested an improvement about it in this PR. I'm just saying that precisely here there's room for interpretation and brainstorming.
That being said, if you insist I switch to your version, because in general I don't like comments. Even mine :)